File: rust-solutions/src/day3.rs
use crate::util::*;
use std::collections::HashSet;

pub fn day3(p: Part) -> usize {
    match p {
        Part::One => solve(parse1),
        Part::Two => solve(parse2),
    }
}

fn get_input() -> String {
    std::fs::read_to_string("./input/day3.txt").unwrap()
}

fn solve<F>(parse: F) -> usize
where
    F: Fn() -> Vec<Vec<String>>,
{
    let string_to_set = |s: &String| -> HashSet<char> {
        let mut hs = HashSet::new();
        s.chars().for_each(|c| {
            hs.insert(c);
        });
        hs
    };

    parse()
        .iter()
        .map(|v| intersect_hashsets(v.iter().map(string_to_set)))
        .map(|hs| value(*hs.iter().next().unwrap())) // ????
        .sum()
}

fn parse1() -> Vec<Vec<String>> {
    get_input()
        .lines()
        .map(|s| {
            let split = s.len() / 2;
            vec![s[0..split].to_string(), s[split..].to_string()]
        })
        .collect()
}

fn parse2() -> Vec<Vec<String>> {
    get_input()
        .lines()
        .map(|s| s.to_string())
        .collect::<Vec<String>>()
        .chunks(3)
        .map(|x| x.to_vec())
        .collect()
}

// How is this so hard/not built into the HashSet module?
fn intersect_hashsets<I>(mut hash_sets: I) -> HashSet<char>
where
    I: Iterator<Item = HashSet<char>>,
{
    if let Some(first) = hash_sets.next() {
        hash_sets.fold(first, |acc, el| {
            acc.intersection(&el).copied().collect() // uff²
        })
    } else {
        HashSet::new()
    }
}

fn value(c: char) -> usize {
    let ascii = c as usize;
    if (97..=122).contains(&ascii) {
        ascii - 96
    } else {
        ascii - 38
    }
}

File: aoc2022/clojure-solutions/src/clojure_solutions/day3.clj
(ns clojure-solutions.day3
  (:require [clojure.string :as str]
            [clojure.set    :as set])
  (:use [clojure-aoc-util.util] :reload))

(defn- parse []
  (str/split-lines (slurp "../inputs/day3.txt")))

(defn- value [c]
  (let [ascii (int c)]
    (cond
      (<= 97 ascii 122) (- ascii 96)
      (<= 65 ascii 90 ) (- ascii 38))))

(defn- solve [fun]
  (->> (parse)                           ; ["AA" "acD" …]
       fun                               ; [[[\A \A] …] [[\c \Z \a] …] …]
       (map #(map (partial into #{}) %)) ; [[#{\A} …] [#{\c \Z \a} …] …]
       (map #(apply set/intersection %)) ; [#{\A} #{\Z} …]
       (map (comp value first))          ; (+ 27 52 …)
       sum))

(defn day3 [part]
  (case part
    :one (solve #(map (fn [s] (split-at (/ (count s) 2) s)) %))
    :two (solve #(partition 3 %))))
